--- libsrc/posix-unsafe_process_primitives.adb	2017-05-16 12:40:58.000000000 +0200
+++ libsrc/posix-unsafe_process_primitives.adb	2018-02-18 13:36:04.826431569 +0100
@@ -89,8 +89,8 @@
       Result : pid_t;
       package SSL renames System.Soft_Links;
       --  save local values of soft-link data
-      NT_Sec_Stack_Addr : constant System.Address :=
-                            SSL.Get_Sec_Stack_Addr.all;
+      NT_Sec_Stack_Addr : constant SSL.SST.SS_Stack_Ptr :=
+                            SSL.Get_Sec_Stack.all;
       NT_Jmpbuf_Address : constant System.Address :=
                             SSL.Get_Jmpbuf_Address.all;
    begin
@@ -106,10 +106,10 @@
          SSL.Unlock_Task        := SSL.Task_Unlock_NT'Access;
          SSL.Get_Jmpbuf_Address := SSL.Get_Jmpbuf_Address_NT'Access;
          SSL.Set_Jmpbuf_Address := SSL.Set_Jmpbuf_Address_NT'Access;
-         SSL.Get_Sec_Stack_Addr := SSL.Get_Sec_Stack_Addr_NT'Access;
-         SSL.Set_Sec_Stack_Addr := SSL.Set_Sec_Stack_Addr_NT'Access;
+         SSL.Get_Sec_Stack      := SSL.Get_Sec_Stack_NT'Access;
+         SSL.Set_Sec_Stack      := SSL.Set_Sec_Stack_NT'Access;
          --  reset global data to saved local values for this thread
-         SSL.Set_Sec_Stack_Addr (NT_Sec_Stack_Addr);
+         SSL.Set_Sec_Stack (NT_Sec_Stack_Addr);
          SSL.Set_Jmpbuf_Address (NT_Jmpbuf_Address);
       end if;
       return To_Process_ID (Result);
